File: apps/web-dashboard/lib/scenarioPlanner/dateCalendar.ts
/** ISO date (YYYY-MM-DD) math in UTC to avoid DST surprises. */

const ISO = /^(\d{4})-(\d{2})-(\d{2})$/;

export function parseIsoDateUTC(iso: string): Date {
  const m = ISO.exec(iso.trim());
  if (!m) throw new Error(`Invalid ISO date: ${iso}`);
  return new Date(Date.UTC(Number(m[1]), Number(m[2]) - 1, Number(m[3])));
}

export function formatIsoDateUTC(d: Date): string {
  const y = d.getUTCFullYear();
  const mo = String(d.getUTCMonth() + 1).padStart(2, '0');
  const day = String(d.getUTCDate()).padStart(2, '0');
  return `${y}-${mo}-${day}`;
}

export function addCalendarDaysUTC(iso: string, deltaDays: number): string {
  const d = parseIsoDateUTC(iso);
  d.setUTCDate(d.getUTCDate() + deltaDays);
  return formatIsoDateUTC(d);
}

/** Inclusive calendar-day count from start through end (same day => 1). */
export function inclusiveCalendarDaySpan(startIso: string, endIso: string): number {
  const s = parseIsoDateUTC(startIso).getTime();
  const e = parseIsoDateUTC(endIso).getTime();
  if (e < s) return 0;
  return Math.round((e - s) / 86400000) + 1;
}

/** Whole-day offset from date a to date b (b - a in calendar days). */
export function calendarDaysOffset(fromIso: string, toIso: string): number {
  const a = parseIsoDateUTC(fromIso).getTime();
  const b = parseIsoDateUTC(toIso).getTime();
  return Math.round((b - a) / 86400000);
}

File: apps/agent-worker/worker/scenario_engine/date_calendar.py
"""ISO YYYY-MM-DD calendar math (UTC-style, no DST). Parity with dateCalendar.ts."""

from __future__ import annotations

import re
from datetime import date, timedelta

_ISO = re.compile(r'^(\d{4})-(\d{2})-(\d{2})$')


def parse_iso_date_utc(iso: str) -> date:
    m = _ISO.match(iso.strip())
    if not m:
        raise ValueError(f'Invalid ISO date: {iso}')
    y, mo, d = int(m.group(1)), int(m.group(2)), int(m.group(3))
    return date(y, mo, d)


def format_iso_date_utc(d: date) -> str:
    return d.isoformat()


def add_calendar_days_utc(iso: str, delta_days: int) -> str:
    d = parse_iso_date_utc(iso)
    return format_iso_date_utc(d + timedelta(days=delta_days))


def inclusive_calendar_day_span(start_iso: str, end_iso: str) -> int:
    s = parse_iso_date_utc(start_iso)
    e = parse_iso_date_utc(end_iso)
    if e < s:
        return 0
    return (e - s).days + 1


def calendar_days_offset(from_iso: str, to_iso: str) -> int:
    a = parse_iso_date_utc(from_iso)
    b = parse_iso_date_utc(to_iso)
    return (b - a).days
